renderdoc/os/posix/linux/linux_stringio.cpp

```diff
@@ -241,6 +245,112 @@ bool GetXCBKeyState(int key)
 
 #endif
 
+#if ENABLED(RDOC_WAYLAND)
+
+struct wl_seat *CurrentWaylandSeat = NULL;
+struct wl_display *CurrentWaylandDisplay = NULL;
+struct wl_keyboard *CurrentWaylandKeyboard = NULL;
+
+static void KeyboardHandleEnter(void *data, struct wl_keyboard *keyboard, uint32_t serial,
+                                struct wl_surface *surface, struct wl_array *keys)
+{
+}
+
+static void KeyboardHandleLeave(void *data, struct wl_keyboard *keyboard, uint32_t serial,
+                                struct wl_surface *surface)
+{
+}
+
+static void KeyboardHandleKey(void *data, struct wl_keyboard *keyboard, uint32_t serial,
+                              uint32_t time, uint32_t key, uint32_t state_w)
+{
+  if(state_w == WL_KEYBOARD_KEY_STATE_PRESSED)
+  {
+    printf("Key %c pressed!\n", key);
+  }
+  else if(state_w == WL_KEYBOARD_KEY_STATE_RELEASED)
+  {
+    printf("Key %c released!\n", key);
+  }
+}
+
+static const struct wl_keyboard_listener KeyboardListener = {
+    NULL, KeyboardHandleEnter, KeyboardHandleLeave, KeyboardHandleKey, NULL, NULL,
+};
+
+static void SeatHandleCapabilities(void *data, struct wl_seat *seat, uint32_t caps)
+{
+  if((caps & WL_SEAT_CAPABILITY_KEYBOARD) && !CurrentWaylandKeyboard)
+  {
+    CurrentWaylandKeyboard = wl_seat_get_keyboard(seat);
+    /* wl_keyboard_set_user_data(input->keyboard, input); */
+    wl_keyboard_add_listener(CurrentWaylandKeyboard, &KeyboardListener, NULL);
+  }
+  else if(!(caps & WL_SEAT_CAPABILITY_KEYBOARD) && CurrentWaylandKeyboard)
+  {
+    wl_keyboard_destroy(CurrentWaylandKeyboard);
+    CurrentWaylandKeyboard = NULL;
+  }
+}
+
+static const struct wl_seat_listener SeatListener = {
+    SeatHandleCapabilities, NULL,
+};
+
+static void RegistryAddObject(void *data, struct wl_registry *registry, uint32_t name,
+                              const char *interface, uint32_t version)
+{
+  if(!strcmp(interface, "wl_seat"))
+  {
+    CurrentWaylandSeat = (struct wl_seat *)wl_registry_bind(registry, name, &wl_seat_interface, 1);
+    wl_seat_add_listener(CurrentWaylandSeat, &SeatListener, NULL);
+  }
+}
+
+static const struct wl_registry_listener RegistryListener = {
+    RegistryAddObject,
+};
+
+void AddWaylandInputWindow(void *wnd)
+{
+  if(wnd == NULL)
+    return;
+
+  if(!CurrentWaylandDisplay)
+    CurrentWaylandDisplay = wl_display_connect(NULL);
+
+  struct wl_registry *registry = wl_display_get_registry(CurrentWaylandDisplay);
```

Comment:
I realise this code isn't called yet, but I had a couple of comments on this.

I assume you'll check whether we're running wayland outside this function, because it assumes we are definitely using wayland by not checking if `wl_display_connect()` succeeded.

Also the code as it stands suggests that we don't have to do anything at all to grab keyboard input globally. We connect to the display just with `NULL`, and don't use the window handle at all. It's not necessarily wrong I'm just checking my assumptions that it works like xlib/xcb in that respect. I'm also assuming there's no problem with opening a second wayland display handle here, on xlib/xcb that could cause driver crashes if we don't consistently use only a single connection.

Lastly there's one other issue - in `gl_hooks_egl.cpp` in the hooked `eglGetDisplay` we call `CloneDisplay()` further up in this file on the native display handle passed, assuming it will be an xlib `Display *`. On wayland I suspect that treating a wayland display handle like an xlib display handle will probably crash. Might it be a good idea to move this code that calls `wl_display_connect()` into that function - then if the wayland connection succeeds we assume the display we were just passed was a wayland display and not an xlib display. Does that make sense?

Reply:
Yeah, this code is certainly not finished. But in response to your questions, yes, we are assuming that we are definitely using Wayland when calling this function since the rest of the codebase is littered with X11-related `ifdef`s and other assumptions, so for the purposes of hacking in basic support, I was making a similar assumption as well.

With Wayland, `wl_display_connect(NULL)` returns a pointer to the currently active `wl_display` and doesn't open a new one. Replacing the `NULL` with a `const char *` will attempt to connect to a different display with the given name. AFAICT, Wayland applications can open as many `wl_display` handles as they want.

Reply:
@baldurk Looks like GitHub decided to hide these comments since I changed this one line from `struct wl_registry *` to just `wl_registry *`. I hope you see this!

Reply:
Yep I see the emails for each comment :smile:.

It sounds like you have a handle on it all, I just wanted to be sure there weren't lurking issues from problems I've run into in the past.

Also note that the `WindowingData` is only used during replay, it isn't part of the code for the injected application at all, and similarly the keyboard input is only used during capture in the injected application for detecting shortcut keypresses. So in that respect it's not related to the code here.